app/controllers/collections_controller.rb

```diff
@@ -189,4 +233,218 @@ def respond_with_error(message, status_code)
+  def upload_document_using_multipart(corpus_dir, file_basename, file, collection_name)
+    absolute_filename = File.join(corpus_dir, file_basename)
+    if !file.is_a? ActionDispatch::Http::UploadedFile
+      raise ResponseError.new(412), "Error in file parameter."
+    elsif file.blank? or file.size == 0
+      raise ResponseError.new(412), "Uploaded file \"#{file_basename}\" is not present or empty."
+    else
+      Rails.logger.debug("Copying uploaded document file from #{file.tempfile} to #{absolute_filename}")
+      FileUtils.cp file.tempfile, absolute_filename
+      absolute_filename
+    end
+  end
+
+  # Uploads any documents in the item metadata and returns a copy of the item metadata with its metadata graph updated
+  def process_item_documents_and_update_graph(corpus_dir, item_metadata)
+    unless item_metadata["documents"].nil?
+      item_metadata["documents"].each do |document|
+        doc_abs_path = upload_document_using_json(corpus_dir, document["identifier"], document["content"])
+        unless doc_abs_path.nil?
+          item_metadata['metadata']['@graph'] = update_document_source_in_graph(item_metadata['metadata']['@graph'], document["identifier"], doc_abs_path)
+        end
+      end
+    end
+    item_metadata
+  end
+
+  # Updates the metadata graph for the given item
+  # Returns a copy of the item with the document sourcs in the graph updates to the path of an uploaded file when appropriate
+  def update_item_graph_with_uploaded_files(uploaded_files, item_metadata)
+    unless uploaded_files.nil?
+      uploaded_files.each do |file_path|
+        item_metadata['metadata']['@graph'] = update_document_source_in_graph(item_metadata['metadata']['@graph'], File.basename(file_path), file_path)
+      end
+    end
+    item_metadata
+  end
+
+  # Updates the source of a document in the JSON formatted Item graph
+  def update_document_source_in_graph(json_graph, document_identifier, document_source)
+    json_graph.each do |graph_entry|
+      if graph_entry['dcterms:identifier'] == document_identifier
+        # Escape any filename spaces with '%20' as URIs with spaces are flagged as invalid when RDF loads
+        graph_entry.update({'dcterms:source' => {'@id' => "file://#{document_source.sub(" ", "%20")}"}})
+        json_graph
+      end
+    end
+  end
+
+  # Checks if json is valid
+  def valid_json?(json)
+    begin
+      JSON.parse(json)
+      true
+    rescue
+      false
+    end
+  end
+
+  # Returns a cleansed copy of params for the add item api
+  def cleanse_params(request_params)
+    unless request_params.nil?
+      if request_params[:items].is_a? String
+        raise ResponseError.new(400), "JSON item metadata is ill-formatted" unless valid_json?(request_params[:items])
+        request_params[:items] = JSON.parse(request_params[:items])
+      end
+      request_params[:file] = [request_params[:file]] unless request_params[:file].nil? or request_params[:file].is_a? Array
```

Comment:
Returning empty collections in general is better than leaving something `nil` because it simplifies iteration code further down that track. Later on you can do:
```
files.each do |file|
...
end
```

instead of:
```
unless files.nil?
  files.each do |file|
  ...
  end
end

```